src/main/filesystem/watcher.js

```diff
@@ -96,61 +126,113 @@ class Watcher {
    */
   constructor (preferences) {
     this._preferences = preferences
+    this._ignoreChangeEvents = []
     this.watchers = {}
   }
 
-  // return a unwatch function
+  // Watch a file or directory and return a unwatch function.
   watch (win, watchPath, type = 'dir'/* file or dir */) {
     const id = getUniqueId()
     const watcher = chokidar.watch(watchPath, {
       ignored: /(^|[/\\])(\..|node_modules)/,
       ignoreInitial: type === 'file',
-      persistent: true
+      persistent: true,
+      ignorePermissionErrors: true,
+
+      // Just to be sure when a file is replaced with a directory don't watch recursively.
+      depth: type === 'file' ? 0 : undefined,
+
+      // Please see GH#1043
+      awaitWriteFinish: {
+        stabilityThreshold: WATCHER_STABILITY_THRESHOLD,
+        pollInterval: WATCHER_STABILITY_POLL_INTERVAL
+      }
     })
 
+    let disposed = false
+    let enospcReached = false
+    let renameTimeout = null
+
     watcher
-      .on('add', pathname => add(win, pathname, this._preferences.getPreferedEOL()))
-      .on('change', pathname => change(win, pathname, type, this._preferences.getPreferedEOL()))
+      .on('add', pathname => {
+        if (!this._shouldIgnoreEvent(win.id, pathname, type)) {
+          add(win, pathname, type, this._preferences.getPreferedEOL())
+        }
+      })
+      .on('change', pathname => {
+        if (!this._shouldIgnoreEvent(win.id, pathname, type)) {
+          change(win, pathname, type, this._preferences.getPreferedEOL())
+        }
+      })
       .on('unlink', pathname => unlink(win, pathname, type))
-      .on('addDir', pathname => addDir(win, pathname))
-      .on('unlinkDir', pathname => unlinkDir(win, pathname))
-      .on('raw', (event, path, details) => {
+      .on('addDir', pathname => addDir(win, pathname, type))
+      .on('unlinkDir', pathname => unlinkDir(win, pathname, type))
+      .on('raw', (event, subpath, details) => {
         if (global.MARKTEXT_DEBUG_VERBOSE >= 3) {
-          console.log(event, path, details)
+          console.log('watcher: ', event, subpath, details)
         }
 
-        // rename syscall on Linux (chokidar#591)
+        // Fix atomic rename on Linux (chokidar#591).
+        // TODO: This should also apply to macOS.
+        // TODO: Do we need to rewatch when the watched directory was renamed?
         if (isLinux && type === 'file' && event === 'rename') {
-          const { watchedPath } = details
-          // Use the same watcher and re-watch the file.
-          watcher.unwatch(watchedPath)
-          watcher.add(watchedPath)
+          clearTimeout(renameTimeout)
```

Comment:
Do you need to ensure `renameTimeout` is not `null` before `clearTimeout`? I also prefer to rename `renameTimeout` to `renameTimer`.

Reply:
I'll rename the variable but clearTimeout on `null` is fine and it requires an id, so null is valid.

Reply:
I mean there is no need to call clearTimeout(Javascript call function consumption) if renameTimeout is null.